src/slice.c

```diff
@@ -191,9 +191,25 @@ static SEXP vec_slice_impl(SEXP x, SEXP index) {
   case vctrs_type_character:
   case vctrs_type_raw:
   case vctrs_type_list: {
-    SEXP out = PROTECT_N(vec_slice_base(info.type, data, index), &nprot);
+    SEXP out;
+
+    if (has_dim(x)) {
+      out = PROTECT_N(vec_slice_shaped(info.type, data, index), &nprot);
+
+      SEXP names = PROTECT_N(Rf_getAttrib(x, R_DimNamesSymbol), &nprot);
+      if (names != R_NilValue) {
+        names = PROTECT_N(Rf_shallow_duplicate(names), &nprot);
+        SEXP row_names = VECTOR_ELT(names, 0);
+        SET_VECTOR_ELT(names, 0, slice_names(row_names, index));
+        Rf_setAttrib(out, R_DimNamesSymbol, names);
+      }
+    } else {
+      out = PROTECT_N(vec_slice_base(info.type, data, index), &nprot);
+
+      SEXP names = PROTECT_N(Rf_getAttrib(x, R_NamesSymbol), &nprot);
+      Rf_setAttrib(out, R_NamesSymbol, slice_names(names, index));
```

Comment:
@lionel- this doesn't need a `PROTECT` right? Since we set the attrib immediately?

Reply:
oh yes it does, allocating the attribute pairlist nodes might trigger a gc.